pkg/agent/openflow/pipeline.go

```diff
@@ -825,32 +966,187 @@ func (c *client) l3ToExternalFlows(nodeIP net.IP, localSubnet net.IPNet, outputP
 	return flows
 }
 
+// serviceLearnFlow generates the flow with learn action which adds new flows in
+// sessionAffinityTable according to the Endpoint selection decision.
+func (c *client) serviceLearnFlow(svcIP net.IP, svcPort uint16, protocol binding.Protocol, affinityTimeout uint16) binding.Flow {
+	learnFlowBuilder := c.pipeline[serviceLBTable].BuildFlow(priorityLow).
+		MatchRegRange(int(serviceLearnReg), marksRegServiceNeedLearn, serviceLearnRegRange).
+		MatchDstIP(svcIP).
+		Cookie(c.cookieAllocator.Request(cookie.Service).Raw())
+	learnFlowBuilderLearnAction := learnFlowBuilder.
+		Action().Learn(sessionAffinityTable, priorityNormal, affinityTimeout, 0, c.cookieAllocator.Request(cookie.Service).Raw()).
+		DeleteLearned()
+	if protocol == binding.ProtocolTCP {
```

Comment:
Note the protocols supported in Kubernetes Service.

Reply:
Done.